commit 2a296b0517bce574d0e880f73572e4806329ed7e (refs/changes/65/965/4)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2016-12-08T14:08:00+13:00 (2 years, 10 months ago)
    
    PUBAPI-1291: primaryIp's logic could be better
    Reviewed by: Trent Mick <trent.mick@joyent.com>

diff --git a/docs/index.md b/docs/index.md
index d497992..f3e1a08 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -3927,7 +3927,7 @@ updated     | ISO8601 date | When this instance's details was last updated
 docker      | Boolean  | Whether this instance is a Docker container, if present
 ips         | Array[String] | The IP addresses this instance has
 networks    | Array[String] | The network UUIDs of the nics this instance has
-primaryIp   | String   | IP address of the primary nic of this instance
+primaryIp   | String   | The IP address of the primary NIC of this instance. The "primary" NIC is used to determine the default gateway for an instance. Commonly it is also on an external network (i.e. accessible on the public internet) and hence usable for SSH'ing into an instance, but not always. (Note: In future Triton versions it will be possible to have multiple IPv4 and IPv6 addresses on a particular NIC, at which point the current definition of `primaryIp` will be ambiguous and will need to change.)
 firewall_enabled | Boolean  | Whether firewall rules are enforced on this instance
 compute_node | String  | UUID of the server on which the instance is located
 package     | String   | The id or name of the package used to create this instance
@@ -4080,11 +4080,11 @@ updated     | ISO8601 date | When this instance's details was last updated
 docker      | Boolean  | Whether this instance is a Docker container, if present
 ips         | Array[String] | The IP addresses this instance has
 networks    | Array[String] | The network UUIDs of the nics this instance has
-primaryIp   | String   | IP address of the primary nic of this instance
+primaryIp   | String   | The IP address of the primary NIC of this instance. The "primary" NIC is used to determine the default gateway for an instance. Commonly it is also on an external network (i.e. accessible on the public internet) and hence usable for SSH'ing into an instance, but not always. (Note: In future Triton versions it will be possible to have multiple IPv4 and IPv6 addresses on a particular NIC, at which point the current definition of `primaryIp` will be ambiguous and will need to change.)
 firewall_enabled | Boolean  | Whether firewall rules are enforced on this instance
 compute_node | String  | UUID of the server on which the instance is located
 package     | String   | The id or name of the package used to create this instance
-dns_names   | Array[String] | DNS names of the instance
+dns_names   | Array[String] | DNS names of the instance (if the instance is using [CNS](https://docs.joyent.com/public-cloud/network/cns))
 
 ### Errors
 
diff --git a/lib/machines.js b/lib/machines.js
index 201dc98..9ab90bf 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -164,17 +164,22 @@ function translate(machine, req)  {
             }
         });
 
-        msg.primaryIp = machine.nics.reduce(function (acc, nic) {
-            return (nic.nic_tag === 'external') ? nic.ip : acc;
-        }, '');
-
-        // PUBAPI-727: If a machine has only internal networks, use that
-        // as the primary IP
-        if (msg.primaryIp === '') {
-            msg.primaryIp = machine.nics.reduce(function (acc, nic) {
-                return (nic.nic_tag === 'internal') ? nic.ip : acc;
-            }, '');
+        var primaryNic = machine.nics.filter(function (nic) {
+            return nic.primary;
+        })[0];
+
+        // For backwards-compat. Some old VMs don't have a primary
+        // attribute set on their nics, but have an 'external' nic tag.
+        if (!primaryNic) {
+            primaryNic = machine.nics.filter(function (nic) {
+                return nic.nic_tag === EXTERNAL_NIC_TAG;
+            }).pop();
         }
+
+        // if we found nothing yet, just pick any NIC
+        primaryNic = primaryNic || machine.nics[0];
+
+        msg.primaryIp = primaryNic ? primaryNic.ip : '';
     }
 
     if (credentials && machine.internal_metadata &&
diff --git a/test/machines/common.js b/test/machines/common.js
index 58a0765..d3fd836 100644
--- a/test/machines/common.js
+++ b/test/machines/common.js
@@ -81,6 +81,10 @@ function checkMachine(t, m) {
     t.ok(m.metadata, 'checkMachine metadata ok');
     t.notOk(m.docker, 'checkMachine docker attr not set');
 
+    if (m.state === 'running') {
+        t.ok(m.primaryIp, 'checkMachine primaryIp ok');
+    }
+
     // Sometimes test suites from other applications create zones with a
     // 00000000-0000-0000-0000-000000000000 billing_id, which is changed by
     // cloudapi to '' since it's not an actual package UUID. Alas, we work
